quark/foreach.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 *--
+
+import os
+import subprocess
+import contextlib
+from argparse import ArgumentParser
+
+from .subproject import Subproject
+from .subproject import SvnSubproject
+
+
+@contextlib.contextmanager
+def change_dir(path):
+    _old_path = os.getcwd()
+
+    try:
+        os.chdir(path)
+        yield
+    finally:
+        os.chdir(_old_path)
+
+
+def run():
+    parser = ArgumentParser(
+        description="""
+        Evaluates an arbitrary shell command in each submodule, skipping all
+        the svn's submodules.
+        The command has access to the variables $name, $sm_path, $displaypath,
+        $sha1 and $toplevel:
+        $name is the name of the submodule;
+        $sm_path is the path of the submodule relative to the superproject;
+        $displaypath is the path of the submodule relative to the root directory;
+        $sha1 is the commit as recorded in the immediate superproject;
+        $toplevel is the absolute path to the top-level of the immediate superproject.
+        """
+    )
+    parser.add_argument(
+        "-q", "--quiet",
+        action="store_true",
+        help="Only print error messages"
+    )
+    parser.add_argument(
+        "command",
+        action="store",
+        nargs="+",
+        help="The command that will be run for every dependency"
+    )
+
+    optlist = parser.parse_args()
+    optlist.source_directory = os.getcwd()
+
+    root, modules = Subproject.create_dependency_tree(
+        optlist.source_directory, update=False
+    )
+
+    for path, module in modules.items():
+        version_control = "svn" if type(module) is SvnSubproject else "git"
+        module_relpath = os.path.relpath(
+            module.directory, optlist.source_directory
+        )
+        module_displaypath = os.path.relpath(
+            module.directory, os.getcwd()
+        )
+        module_name = module.name
+        toplevel = os.path.abspath(optlist.source_directory)
+
+        if version_control == "git":
+            # Case for git
+            commit_sha = module.ref
+        else:
+            # Case for SVN
+            if not optlist.quiet:
+                print(
+                    "WARNING: {} at {} has been skipped because it's a "
+                    "SVN repository\n".format(
+                        module_name,
+                        module_displaypath
+                    ))
+            continue
```

Comment:
Consider adding a varaible similar to "sha1" and containing the revision number for SVN. If it isn't feasable I would just not set it and set all the other ones.

Reply:
Solved in https://github.com/comelz/quark/pull/27/commits/ac161eb06068a77e0ed47763b93be0bf38ce804e